### PR TITLE
tests: Add testing code to cover all code in the UserInfoProvider (SDKCF-6380)

### PR DESCRIPTION
**Pull request by ts-robih-robihamanto**
# Description
Increase code coverage in UserInfoProvider.swift
Include tests to cover `!=` when comparing the object.

## Links
SDKCF-6380

# Checklist
- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have added to the [changelog](../blob/master/CHANGELOG.md#Unreleased)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data **before every commit**, including API endpoints and keys, and internal links
- [ ] I ran `fastlane ci` without errors
- [x] All project file changes are replicated in `SampleSPM/SampleSPM.xcodeproj` project

### CODE REVIEW

File: Sources/RInAppMessaging/UserInfoProvider.swift [hunk 1]
```diff
@@ -66,8 +66,14 @@ func == (lhs: UserInfoProvider?, rhs: UserInfoProvider?) -> Bool {
 
 @inlinable
 func != (lhs: UserInfoProvider?, rhs: UserInfoProvider?) -> Bool {
-    let equalsResult = lhs == rhs
-    return !equalsResult
+    switch (lhs, rhs) {
```
Comment (imaginaris) [inline Sources/RInAppMessaging/UserInfoProvider.swift:69]: Was this change necessary?

File: Tests/Tests/UserInfoProviderSpec.swift [hunk 2]
```diff
@@ -157,6 +157,106 @@ class UserInfoProviderSpec: QuickSpec {
...
+            context("when checking inequality") {
+                var userInfoProviderA: UserInfoProviderMock!
+                var userInfoProviderB: UserInfoProviderMock!
+
+                beforeEach {
+                    userInfoProviderA = UserInfoProviderMock()
+                    userInfoProviderB = UserInfoProviderMock()
+                }
+
+                it("will return true for nil object") {
```
Comment (imaginaris) [inline Tests/Tests/UserInfoProviderSpec.swift]: I think you meant "will return false..."

File: Tests/Tests/UserInfoProviderSpec.swift [hunk 3]
```diff
@@ -157,6 +157,106 @@ class UserInfoProviderSpec: QuickSpec {
...
+                    userInfoProviderB = nil
+                    expect(userInfoProviderA != userInfoProviderB).to(beFalse())
+                }
+
+                it("will return true for nil and empty objects") {
+                    userInfoProviderA = nil
+                    expect(userInfoProviderA != userInfoProviderB).to(beTrue())
+                }
+
+                it("will return false for different userIds") {
```
Comment (imaginaris) [inline Tests/Tests/UserInfoProviderSpec.swift:181]: But userIds are not different in this test

**Review by imaginaris — CHANGES_REQUESTED** [review]

File: Sources/RInAppMessaging/UserInfoProvider.swift (hunk 1 above)
Comment (ts-robih-robihamanto) [reply Sources/RInAppMessaging/UserInfoProvider.swift:69]: Yes, otherwise the code on 87 - 91 will not get called:
```
@inlinable
func != (lhs: UserInfoProvider, rhs: UserInfoProvider) -> Bool {
    let equalsResult = lhs == rhs
    return !equalsResult
}
```

File: Tests/Tests/UserInfoProviderSpec.swift (hunk 3 above)
Comment (ts-robih-robihamanto) [reply Tests/Tests/UserInfoProviderSpec.swift:181]: Oh ya, the user info is treated by userIds in this case should return false.

Comment (ts-robih-robihamanto) [reply Tests/Tests/UserInfoProviderSpec.swift:181]: Fixed here [8b78146](https://github.com/rakutentech/ios-inappmessaging/pull/257/commits/a88a42758571ac69c74129b85cc4835ee8b78146).

File: Tests/Tests/UserInfoProviderSpec.swift (hunk 2 above)
Comment (ts-robih-robihamanto) [reply Tests/Tests/UserInfoProviderSpec.swift]: Fixed here [e4ab56e](https://github.com/rakutentech/ios-inappmessaging/pull/257/commits/e9822a7e2d95d60d8837db3f031f2b990e4ab56e).

File: Sources/RInAppMessaging/UserInfoProvider.swift (hunk 1 above)
Comment (imaginaris) [reply Sources/RInAppMessaging/UserInfoProvider.swift:69]: Hmm. But it would if you used non optional `UserInfoProviderMock` objects in your test, wouldn't it?

File: Tests/Tests/UserInfoProviderSpec.swift [hunk 4]
```diff
@@ -157,6 +157,106 @@ class UserInfoProviderSpec: QuickSpec {
...
+
+                it("will return false for nil object") {
+                    userInfoProviderA = nil
+                    userInfoProviderB = nil
+                    expect(userInfoProviderA != userInfoProviderB).to(beFalse())
+                }
+
+                it("will return false for nil and empty objects") {
+                    userInfoProviderA = nil
+                    expect(userInfoProviderA != userInfoProviderB).to(beFalse())
```
Comment (imaginaris) [inline Tests/Tests/UserInfoProviderSpec.swift:178]: so empty object and nil are treated as equal? 🤔 It doesn't sound correct to me

File: Sources/RInAppMessaging/UserInfoProvider.swift (hunk 1 above)
Comment (imaginaris) [reply Sources/RInAppMessaging/UserInfoProvider.swift:69]: but maybe it's better to keep it your way after all

File: Tests/Tests/UserInfoProviderSpec.swift (hunk 4 above)
Comment (ts-robih-robihamanto) [reply Tests/Tests/UserInfoProviderSpec.swift:178]: Yes, I think because both of the `userIds ` is empty it's equal.
In the UserInfoProvider | L.81 you made a comment:
**// null and empty values are treated the same**
So, it should be correct, right?

Comment (imaginaris) [reply Tests/Tests/UserInfoProviderSpec.swift:178]: Ahh ok. That makes sense. I remember now 😅. Nil and empty user are treated as the same to improve 'logout' state detection. The test is absolutely correct.

**Review by imaginaris — APPROVED** [review]
